File: src/chapters/nonlinear-equations/content/code/fixed-point-nd.r
fixed_point_nd <- function(G, x0, tol = 1e-12, max_iter = 200) {
  # Vector fixed-point iteration x_{k+1} = G(x_k).
  x <- as.numeric(x0)
  for (k in 1:max_iter) {
    xn <- as.numeric(G(x))
    if (max(abs(xn - x)) < tol) return(list(root = xn, iters = k))
    x <- xn
  }
  list(root = x, iters = max_iter)
}

# x = (cos y, sin x) has a fixed point near (0.768, 0.695)
G <- function(v) c(cos(v[2]), sin(v[1]))
res <- fixed_point_nd(G, c(0.0, 0.0))
cat(sprintf("(array([%.8f, %.8f]), %d)\n", res$root[1], res$root[2], res$iters))

File: src/chapters/nonlinear-equations/content/code/fixed-point-nd.py
import numpy as np


def fixed_point_nd(G, x0, tol=1e-12, max_iter=200):
    """Vector fixed-point iteration x_{k+1} = G(x_k)."""
    x = np.array(x0, float)
    for k in range(1, max_iter + 1):
        xn = np.array(G(x), float)
        if np.linalg.norm(xn - x, np.inf) < tol:
            return xn, k
        x = xn
    return x, max_iter


if __name__ == "__main__":
    # x = (cos y, sin x) has a fixed point near (0.694, 0.640)
    G = lambda v: [np.cos(v[1]), np.sin(v[0])]
    print(fixed_point_nd(G, [0.0, 0.0]))
